File: pnad.php
<?php

	ini_set('max_execution_time', 2000); //240 seconds = 4 minutes

	// Calculate time
	$time_start = microtime(true);

	// Read Column mapping from SAS
	$sas_file = "sas/Input PNADC_1Tri_2012 a 3Tri_2015.sas";
	$sas_handle = fopen($sas_file, "r");
	$sas_columns = [];

	if ($sas_handle) {
	    while (($line = fgets($sas_handle)) !== false) {
	        // Checks if is an SAS instruction line
	        if (substr($line, 0, 1) == "@") {

		        $line = preg_replace('!\s+!', ' ', $line);
		        
				$tmp_columns = array_slice(str_getcsv($line, " "), 0, 3);
	        	$tmp_columns[0] = intval(str_replace("@", "", $tmp_columns[0]));
				$tmp_columns[1] = utf8_encode($tmp_columns[1]);
				$tmp_columns[2] = intval(str_replace("$", "", str_getcsv( $tmp_columns[2], ".")[0]));

				array_push ($sas_columns, $tmp_columns);
	        }
	    }

	    fclose($sas_handle);
	} else {
	    // error opening the file.
	}


	$end_sas_columns = end($sas_columns);

	// Create sql insert header
	$table = "pnad_trimensal";
	$sql_insert_pnad_header = "INSERT INTO `" . $table . "` (";

	foreach ($sas_columns as $key => $column) {
		$sql_insert_pnad_header .= "`". $column[1] ."`";

		if ($column === $end_sas_columns) {
			$sql_insert_pnad_header .= ") VALUES (";
		} else {
			$sql_insert_pnad_header .= ",";			
		}
        	
	}

	// Read data file
	$data_file = "data/PNADC_012012.txt";
	$data_handle = fopen($data_file, "r");
	$data_lines = 0;

	// Db connection
	$servername = "localhost";
	$username = "root";
	$password = "";
	$dbname = "test";

	// Create connection
	$conn = mysqli_connect($servername, $username, $password, $dbname);
	// Check connection
	if (!$conn) {
	    die("Connection failed: " . mysqli_connect_error());
	}
	// turn OFF auto
	mysqli_autocommit($conn, FALSE); 

	if ($data_handle) {

	    while (($line = fgets($data_handle)) !== false) {
	        // Checks if is an SAS instruction line
	        $sql_insert_pnad_row = "";
			foreach ($sas_columns as $key => $column) {
				
				$sql_insert_pnad_row .= "'" . trim( substr($line, $column[0] - 1, $column[2]) ) . "'";
				
				if ($column === $end_sas_columns) {
					$sql_insert_pnad_row .= ")";
				} else {
					$sql_insert_pnad_row .= ",";			
				}
			}
			// Insert here!
			mysqli_query($conn, $sql_insert_pnad_header . $sql_insert_pnad_row);
			
			$data_lines++;
	    }

	    // Commit your changes
	    mysqli_commit($conn);
	    // Close Database
    	mysqli_close($conn);

	    fclose($data_handle);

	} else {
	    // error opening the file.
	}

	// Execution time in minutes
	$execution_time = (microtime(true) - $time_start);

	// Execution time of the script
	echo '<b>Total Execution Time:</b> ' . $execution_time . ' secs <br>';
	echo '<b>Processed lines:</b> ' . $data_lines. '<br>';
	echo '<b>Sample Line:</b> <br>';

	echo '<pre>';
	echo $sql_insert_pnad_header . $sql_insert_pnad_row;
	echo '</pre>';
?>

File: pnad.py
#!/Python27/python
import traceback, sys
import re, string
import time
import MySQLdb

print "Content-type: text/html"
print
print "<html>"
print "<head>"
print "<title>PNAD</title>"
print "</head>"
print "<body>"

try:
	# Calculate time
	time_start = time.time()

	# Read Column mapping from SAS
	sas_file = "sas/Input PNADC_1Tri_2012 a 3Tri_2015.sas";
	sas_columns = [];

	with open(sas_file) as f:
	    lines = [line.rstrip('\n') for line in f]

	for line in lines:
		# Checks if is an SAS instruction line
		if line[:1] == '@':
	   		line = re.sub(r"\s+"," ",line) 
			tmp_columns =  line.split(" ", 3)
			tmp_columns[0] = int(re.sub("@", "", tmp_columns[0]))-1
	   		tmp_columns[2] = int(string.replace(tmp_columns[2].split(".")[0], '$', ''))
	   		sas_columns.append(tmp_columns)
	
	# Create sql insert header
	table = "pnad_trimensal"
	sql_insert_pnad_header = "INSERT INTO `" + table + "` ("

	for column in sas_columns:
		sql_insert_pnad_header = sql_insert_pnad_header + "`" + column[1]  + "`"
		if column == sas_columns[-1]:
			sql_insert_pnad_header += ") VALUES ("
		else:
			sql_insert_pnad_header += ",";		

	# Create COnnection
	db=MySQLdb.connect(host="localhost",user="root",passwd="",db="test", connect_timeout=2000)
	c=db.cursor()

	# Read data file
	data_file = "data/PNADC_012012.txt"
	data_lines = 0

	with open(data_file) as f:
	    lines = [line.rstrip('\n') for line in f]

	for line in lines:
		sql_insert_pnad_row = ""
		for column in sas_columns:
			sql_insert_pnad_row += "'" + line[column[0]:column[0]+column[2]].strip() + "'";
			if column == sas_columns[-1]:
				sql_insert_pnad_row += ")"
			else:
				sql_insert_pnad_row += ","
		#Insert here!
		c.execute(sql_insert_pnad_header + sql_insert_pnad_row)
		data_lines += 1
	
	# Commit your changes
	db.commit()
	# Close Database
	db.close()

	# Execution time in minutes
	execution_time = time.time() - time_start

	# Execution time of the script
	print '<b>Total Execution Time:</b> ' + str(execution_time) + ' secs <br>'
	print '<b>Processed lines:</b> ' + str(data_lines) + '<br>'
	print '<b>Sample Line:</b> <br>'
	print '<pre>'
	print sql_insert_pnad_header + sql_insert_pnad_row
	print '</pre>'
except:
    print "<pre>{0}</pre>".format(traceback.format_exc())

print "</body>"
print "</html>"